src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -1625,7 +1640,16 @@ private static bool QueueUserWorkItemHelper(WaitCallback callBack, Object state,
                 try { }
                 finally
                 {
-                    QueueUserWorkItemCallback tpcallBack = new QueueUserWorkItemCallback(callBack, state, compressStack, ref stackMark);
+                    ExecutionContext context = compressStack && !ExecutionContext.IsFlowSuppressed() ?
+                                                ExecutionContext.Capture(ref stackMark, ExecutionContext.CaptureOptions.IgnoreSyncCtx | ExecutionContext.CaptureOptions.OptimizeDefaultCase) :
+                                                ExecutionContext.Default;
+
+                    QueueUserWorkItemCallback tpcallBack = (context == ExecutionContext.Default) ?
+                                                (state == null ? 
+                                                    new QueueUserWorkItemCallback(callBack) : // default or null context, no state 
+                                                    new QueueUserWorkItemCallbackWithState(callBack, state)) : // default or null context
```

Comment:
> it shouldn't happen so no need for another class.

It won't happen with QueueUserWorkItem in CoreCLR.  So while the null case still needs to be handled as I outlined, it doesn't need to be optimized.   The key case to optimize is callback+state+ExecutionContext.Default; everything else is secondary.